server/core_purchase.go

```diff
@@ -0,0 +1,153 @@
+	if err != nil {
+		multiLogger.Warn("Skip initialising Google in-app purchase provider", zap.Error(err))
+	}
+
+	return &PurchaseService{
+		logger:       jsonLogger,
+		db:           db,
+		AppleClient:  ac,
+		GoogleClient: gc,
+	}
+}
+
+func (p *PurchaseService) ValidateApplePurchase(userID uuid.UUID, purchase *iap.ApplePurchase) *iap.PurchaseVerifyResponse {
+	r, appleReceipt := p.AppleClient.Verify(purchase)
+	if !r.Success {
+		return r
+	}
+
+	//TODO: Improvement - Process more than one in-app receipts
+	inAppReceipt := appleReceipt.InApp[0]
+	p.checkUser(userID, r, 1, inAppReceipt.TransactionID)
+
+	if r.Success && !r.SeenBefore {
+		err := p.savePurchase(userID.Bytes(), 1, inAppReceipt.ProductID, inAppReceipt.TransactionID, purchase.ReceiptData, r.Data)
+		if err != nil {
+			jsonPurchase, _ := json.Marshal(purchase)
+			p.logger.Error("Could not save Apple purchase", zap.String("receipt", string(jsonPurchase)), zap.String("provider_resp", r.Data), zap.Error(err))
+		}
+	}
+	return r
+}
+
+func (p *PurchaseService) ValidateGooglePurchaseProduct(userID uuid.UUID, purchase *iap.GooglePurchase) *iap.PurchaseVerifyResponse {
+	r, _ := p.GoogleClient.VerifyProduct(purchase)
+	if !r.Success {
+		return r
+	}
+
+	p.checkUser(userID, r, 0, purchase.PurchaseToken)
+	if r.Success && !r.SeenBefore {
+		jsonPurchase, _ := json.Marshal(purchase)
+		err := p.savePurchase(userID.Bytes(), 0, purchase.ProductId, purchase.PurchaseToken, string(jsonPurchase), r.Data)
+		if err != nil {
+			p.logger.Error("Could not save Google product purchase", zap.String("receipt", string(jsonPurchase)), zap.String("provider_resp", r.Data), zap.Error(err))
+		}
+	}
+	return r
+}
+
+func (p *PurchaseService) ValidateGooglePurchaseSubscription(userID uuid.UUID, purchase *iap.GooglePurchase) *iap.PurchaseVerifyResponse {
+	r, _ := p.GoogleClient.VerifySubscription(purchase)
+	if !r.Success {
+		return r
+	}
+
+	p.checkUser(userID, r, 0, purchase.PurchaseToken)
+	if r.Success && !r.SeenBefore {
+		jsonPurchase, _ := json.Marshal(purchase)
+		err := p.savePurchase(userID.Bytes(), 0, purchase.ProductId, purchase.PurchaseToken, string(jsonPurchase), r.Data)
+		if err != nil {
+			p.logger.Error("Could not save Google subscription purchase", zap.String("receipt", string(jsonPurchase)), zap.String("provider_resp", r.Data), zap.Error(err))
+		}
+	}
+	return r
+}
+
+func (p *PurchaseService) checkUser(userID uuid.UUID, r *iap.PurchaseVerifyResponse, provider int, receiptID string) {
+	var purchaseUserID []byte
+	err := p.db.QueryRow("SELECT user_id FROM purchase WHERE provider = $1 AND receipt_id = $2", provider, receiptID).Scan(&purchaseUserID)
+	if err != nil {
+		if err != sql.ErrNoRows {
+			r.Success = false
+			r.Message = errors.New("Failed to validate Apple purchase against ledger.")
+			p.logger.Error(r.Message.Error(), zap.Error(err))
+		}
+	}
+
+	// We've not seen this transaction
+	if purchaseUserID == nil || len(purchaseUserID) == 0 {
```

Comment:
The length of a `nil` slice is 0, so this can be simplified to just `if len(purchaseUserID) == 0 {`.